Add a new setting 'prefix' to decode all parameters starting with it

mcanes opened the pull request:
It can be useful to decode parameters from urls.
Right now, splitting url parameters it's extremely easy using the KV filter, but not decode the parameters values.

Let's put an example using: foo.com/bar?param1=value1&...&paramN=valueN

filters {
  ... # Split url using grok URI pattern or similar

   kv {
     source => "uriparams"
     field_split => "&?"
     prefix => "params_"
   }

   urldecode {
     field_prefix => "params_"
   }
}
With this we can url decode all params using one filter and supporting url params dynamically.

Thanks.

Fixes #2

mcanes committed:
Add a new setting 'prefix' to urldecode all parameters starting with it.

It can be useful to decode all parameters from urls.

mcanes committed:
Fix typo thel -> the

mcanes commented:
Since we need to iterate over all the param values and check the name, it's not very efficient.
I don't know if it can be improved.

purbon commented:
Good morning (Bon dia!) @mcanes, thanks a lot for your contribution!. There are a couple of things need to workout before we can move forward.

First one is you need to sign the CLA (https://www.elastic.co/contributor-agreement), keep in mind to use the same email as the one used to submit this PR, this is required in order to validate for example if you did that at work, or in your personal time.

Second, it would be required to add test to validate your improvement. To do that, you can for example just follow the current tests found at https://github.com/logstash-plugins/logstash-filter-urldecode/blob/master/spec/filters/urldecode_spec.rb, or also write your own in a more rspec friendly way. 

Looking forward to it,

/purbon

mcanes committed:
Add tests to check the prefix setting and the field setting

mcanes commented:
Hi @purbon! Thanks for your help.
I added tests to the new setting and also added one test to the field setting, since as I see, that feature doesn't have any tests.
CLA signed.

Thank you

purbon commented:
@mcanes good work Miquel, reviewing it.

purbon commented on lib/logstash/filters/urldecode.rb:
```diff
@@ -14,7 +14,10 @@ class LogStash::Filters::Urldecode < LogStash::Filters::Base
 
-  # Thel character encoding used in this filter. Examples include `UTF-8`
+  # Urldecode all fields starting with prefix
+  config :prefix, :validate => :string
```
I see a potential confusion with this two config options, while I like the idea of having this kind if filter, not sure what is the best option to configure it. I see different options.
- We change the `all_fields`  to be `string` so it can include a pattern, but then not sure how to properly express the all fields selection, might be with `*`. This would be a breaking change.
- We keep it like is now, but we should add probably a check to see when boths options are setup, or add a warning to make the user aware of setting up `all_fields` and `prefix`. This would be probably the best and easy path for now.

what do you think?

purbon commented:
I added an small comment at https://github.com/logstash-plugins/logstash-filter-urldecode/pull/3/files#r41650714, I think as soon as we set this up, we can safely merge it. Code looks good and test pass.

purbon commented on lib/logstash/filters/urldecode.rb:
```diff
@@ -34,6 +37,13 @@ def filter(event)
       event.to_hash.each { |name, value| event[name] = urldecode(value) }
+    # If a prefix is defined then try to decode all params starting with it
+    elsif !@prefix.nil?
+      event.to_hash.each do |name, value|
```
I recommend  dry this a bit by using the select function (http://ruby-doc.org/core-2.2.2/Hash.html#method-i-select) of a hash. This would make the two options looks alike by doing in case of having the prefix, run the select and filter the hash and then just run the urldecode on the resulting fields.

purbon commented on lib/logstash/filters/urldecode.rb:
```diff
@@ -34,6 +37,13 @@ def filter(event)
+    # If a prefix is defined then try to decode all params starting with it
+    elsif !@prefix.nil?
+      event.to_hash.each do |name, value|
+        if name.start_with?(@prefix)
```
if using select, there is no need to filter at this level.

purbon commented:
don't forget to:
- raise the version accordingly.
- add yourself into the contributors list.

ghost commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

purbon commented:
please jenkins, test this.

purbon commented:
@mcanes Looking forward to have this merged, let me know if you can do the last changes we spoke.

jsvd commented:
there hasn't been any activity in a long time for this PR. @purbon are the review points you made enough to block the merge? I leave it up to you

mcanes commented:
Hi!
Sorry for leaving this pull request without answer it. I'm going to try to finish it asap with the @purbon suggestions.

adamkulcsar commented:
Hi!
Any progress on this issue since May?
